fix(polys): fix DMP.cancel for GF(p) with Flint

oscarbenjamin opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Fixes gh-26804

#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

oscarbenjamin committed:
fix(polys): fix DMP.cancel for GF(p) with Flint

sylee957 commented on sympy/polys/polyclasses.py:
```diff
@@ -2138,19 +2138,28 @@ def _lcm(f, g):
 
-        cH = cF.gcd(cG)
-        cF, cG = cF // cH, cG // cH
+        if R.is_ZZ or R.is_QQ:
```
Is it necessary to check `is_ZZ or is_QQ` here? `is_FiniteField` is narrowed down from above

oscarbenjamin replied:
Oh, good point

oscarbenjamin committed:
fix(polys): remove redundant check

sylee957 commented:
It may be useful to check if field can follow same procedure as `QQ` and can use `clear_denoms` 
But I found some other error in the implementation

```python3
from sympy import *

x = Symbol('x')
Poly(x**2 - 1, modulus=2).clear_denoms()
```
`AttributeError: 'flint.types.nmod_poly.nmod_poly' object has no attribute 'denom'`

oscarbenjamin committed:
fix(polys): various fixes and improvements in DUP_Flint

oscarbenjamin committed:
test(polys): test Poly.clear_denoms with modulus argument

oscarbenjamin commented:
> It may be useful to check if field can follow same procedure as `QQ` and can use `clear_denoms`

The part that is different for `GF(p)` is that we don't normalise the sign of the leading coefficients.

> But I found some other error in the implementation

Thanks. I've fixed that and been through all of the implementation. I found some places where other improvements could be made e.g. to avoid converting to `DMP_Python`.

I found and fixed one other bug:
```python
In [1]: Poly(x, x, modulus=10**100)
---------------------------------------------------------------------------
TypeError
```

sylee957 commented:
I notice another error with factoring over non-prime modulus.
```
(sympy-dev-py3.11) sylee957@DESKTOP-2LJUD4V:~/documents/sympy-dev$ python
Python 3.11.2 (main, Mar 21 2023, 16:28:24) [GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import *
>>> x = Symbol('x')
>>> Poly(x**2 - 1, modulus=4).factor_list()
Flint exception (Impossible inverse):
    Cannot invert modulo 2*2
Aborted
```

Although this should be an error, 
however, I think that the situation is different that this leads to crash of python
that is not possible to recover with try-catch statement.

oscarbenjamin commented:
Actually the simplest solution for now would be to not use flint for non-prime modulus.

oscarbenjamin committed:
fix(polys): Only use Flint for GF(p) if p is prime

python-flint often crashes for operations with nmod when the modulus is
not prime. It is better to raise an exception insteaed. Performance of
GF(p) for non-prime p is not so important (SymPy itself never uses
non-prime p) so for simplicity we don't use Flint if the modulus is not
prime.

Ultimately it would probably be best if GF(p) for non prime p simply
raised an exception when the domain is created.

oscarbenjamin commented:
I believe I have fixed the issues mentioned above. This is ready for review.

sylee957 commented:
Thanks
